core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala

```diff
@@ -284,4 +290,63 @@ class InvokerReactive(
       })
   }
 
+  val healthScheduler = Scheduler.scheduleWaitAtMost(1.seconds)(() => {
+    producer.send("health", PingMessage(instance)).andThen {
+      case Failure(t) => logging.error(this, s"failed to ping the controller: $t")
+    }
+  })
+
+  /** Polls the pool's status and returns a future which completes once the pool is idle. */
+  def waitForContainerPoolIdle(pool: ActorRef): Future[Unit] = {
+    implicit val timeout = Timeout(5 seconds)
+    val delay = 1.second
+
+    (pool ? Busy)
+      .mapTo[Boolean]
+      .flatMap {
+        case true  =>
+          logging.info(this, "Container pool is not idle.")
+          after(delay, actorSystem.scheduler)(waitForContainerPoolIdle(pool))
+        case false =>
+          Future.successful(())
+      }
+      .recoverWith { case _ => after(delay, actorSystem.scheduler)(waitForContainerPoolIdle(pool)) }
+  }
+
+  private val termSignal = "TERM"
+
+  Signal.handle(
+    new Signal(termSignal),
+    new SignalHandler() {
+      override def handle(signal: Signal) = {
+        signal.getName match {
+          case `termSignal` =>
```

Comment:
Is there even a chance that a different signal makes it here since the handler is only attached to a specific signal? could possibly drop all the matching.